GPUP WebGL spends too much time allocating IOSurfaces

kkinnunen-apple opened the pull request:
#### 3b9112ff64a6aa1c0daf62390d1646bcc36b488b
<pre>
GPUP WebGL spends too much time allocating IOSurfaces
<a href="https://bugs.webkit.org/show_bug.cgi?id=255860">https://bugs.webkit.org/show_bug.cgi?id=255860</a>
rdar://108444431

Reviewed by Simon Fraser.

Creating MachSendRight out of IOSurface guarantees that the surface
is marked as in use for the lifetime of the surface. Trust this to
recycle the display buffers as drawing buffers in the GPUP mode.

Increase the held buffers to 3 surfaces per context. Otherwise the
buffers are never recycled, rather always reallocated. The new frame
is always being drawn while the previous frame is still being
composited.

Removes GraphicsContextGLIOSurfaceSwapChain, it is not a useful
abstraction. This was used temporarily when all the OpenGL variants
used the same implementation files.

Renames incorrect naming &quot;DisplayBuffer&quot; to &quot;DrawingBuffer&quot;. The code
actually meant &quot;current drawing buffer that will be display buffer
in the future&quot;.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::reshapeFBOs):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::~GraphicsContextGLCocoa):
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLCocoa::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLCocoa::setDrawingBufferColorSpace):
(WebCore::GraphicsContextGLCocoa::displayBuffer):
(WebCore::GraphicsContextGLCocoa::drawingBuffer):
(WebCore::GraphicsContextGLCocoa::bindNextDrawingBuffer):
(WebCore::GraphicsContextGLCocoa::freeDrawingBuffers):
(WebCore::GraphicsContextGLCocoa::prepareForDisplay):
(WebCore::GraphicsContextGLCocoa::readCompositedResults):
(WebCore::GraphicsContextGLCocoa::paintCompositedResultsToVideoFrame):
(WebCore::GraphicsContextGLCocoa::withDrawingBufferAsNativeImage):
(WebCore::GraphicsContextGLCocoa::withDisplayBufferAsNativeImage):
(WebCore::GraphicsContextGLCocoa::markDisplayBufferInUse): Deleted.
(WebCore::GraphicsContextGLCocoa::reshapeDisplayBufferBacking): Deleted.
(WebCore::GraphicsContextGLCocoa::allocateAndBindDisplayBufferBacking): Deleted.
(WebCore::GraphicsContextGLCocoa::bindDisplayBufferBacking): Deleted.
(WebCore::GraphicsContextGLANGLE::readCompositedResults): Deleted.
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLIOSurfaceSwapChain.cpp: Removed.
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLIOSurfaceSwapChain.h: Removed.
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.cpp:
(WebCore::GraphicsContextGLFallback::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLFallback::reshapeDisplayBufferBacking): Deleted.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.h:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLGBM::reshapeDisplayBufferBacking): Deleted.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLTextureMapperANGLE::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLTextureMapperANGLE::reshapeDisplayBufferBacking): Deleted.
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::prepareForDisplay):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/263496@main">https://commits.webkit.org/263496@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76d5682ccdbe74cdd8b09989393729b7412703a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4784 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5068 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6294 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4912 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5060 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4873 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5154 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4281 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6306 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2431 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4276 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9252 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4293 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4351 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5930 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4759 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3874 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4271 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1173 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8330 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4631 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
```diff
@@ -123,6 +135,9 @@ class WEBCORE_EXPORT GraphicsContextGLCocoa : public GraphicsContextGLANGLE {
 
+    static constexpr size_t maxReusedDrawingBuffers { 3 };
+    size_t m_currentDrawingBufferIndex { 0 };
+    IOSurfacePbuffer m_drawingBuffers[maxReusedDrawingBuffers];
```
Personally, I would have keep this state and `IOSurfacePbuffer` and its related functions in a `swap chain` abstraction.

kkinnunen-apple replied:
I think that turned out to be problematic, as it cannot encapsulate the state. The pbuffer stuff is owned by the context. This created the situation that the swapchain could not manage the surface list by itself, rather the context would still need to manage it and clear it on destruction, etc.

djg reviewed (approved):
:shipit:

Still requires approval from reviewer.

squelart commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -546,83 +535,101 @@ static bool needsEAGLOnMac()
-    }
+GraphicsContextGLCocoa::IOSurfacePbuffer& GraphicsContextGLCocoa::displayBuffer()
+{
+    return m_drawingBuffers[(m_currentDrawingBufferIndex - 1u) % maxReusedDrawingBuffers];
```
If/when `m_currentDrawingBufferIndex` is 0, subtracting 1 makes it roll over to the max value, and modulo a non-power-of-2 doesn't guarantee that it's the wanted 1-step-before.
In this case, (0u-1u)%3==0, which happens to be the same as (1u-1u)%3!
If you're certain that `m_currentDrawingBufferIndex` will actually start at 1u (e.g., because `bindNextDrawingBuffer()` will always be called first?), then it's fine -- though a bit scary, so maybe a comment or `assert(m_c != 0)` would be welcome.
Otherwise you'll want to tweak the formula, e.g.: `m_drawingBuffers[(m_currentDrawingBufferIndex + maxReusedDrawingBuffers - 1u) % maxReusedDrawingBuffers]`.

kkinnunen-apple replied:
@squelart nice catch!
_[Goes to stand in a corner embarrassed for such a fundamental misunderstanding]_

smfr commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
```diff
@@ -105,9 +106,20 @@ class WEBCORE_EXPORT GraphicsContextGLCocoa : public GraphicsContextGLANGLE {
+        void* pbuffer { nullptr };
+        operator bool() const { return !!surface; }
+    };
+    IOSurfacePbuffer& drawingBuffer();
```
can this be `const` ?

kkinnunen-apple replied:
No, the returned ref is non-const by design and the values are stored in the object itself.

smfr commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
```diff
@@ -105,9 +106,20 @@ class WEBCORE_EXPORT GraphicsContextGLCocoa : public GraphicsContextGLANGLE {
+        operator bool() const { return !!surface; }
+    };
+    IOSurfacePbuffer& drawingBuffer();
+    IOSurfacePbuffer& displayBuffer();
```
Ditto

smfr commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
```diff
@@ -123,6 +135,9 @@ class WEBCORE_EXPORT GraphicsContextGLCocoa : public GraphicsContextGLANGLE {
 
+    static constexpr size_t maxReusedDrawingBuffers { 3 };
+    size_t m_currentDrawingBufferIndex { 0 };
+    IOSurfacePbuffer m_drawingBuffers[maxReusedDrawingBuffers];
```
Use `std::array<>` here?

kkinnunen-apple committed:
GPUP WebGL spends too much time allocating IOSurfaces
https://bugs.webkit.org/show_bug.cgi?id=255860
rdar://108444431

Reviewed by Simon Fraser.

Creating MachSendRight out of IOSurface guarantees that the surface
is marked as in use for the lifetime of the surface. Trust this to
recycle the display buffers as drawing buffers in the GPUP mode.

Increase the held buffers to 3 surfaces per context. Otherwise the
buffers are never recycled, rather always reallocated. The new frame
is always being drawn while the previous frame is still being
composited.

Removes GraphicsContextGLIOSurfaceSwapChain, it is not a useful
abstraction. This was used temporarily when all the OpenGL variants
used the same implementation files.

Renames incorrect naming "DisplayBuffer" to "DrawingBuffer". The code
actually meant "current drawing buffer that will be display buffer
in the future".

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::reshapeFBOs):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::~GraphicsContextGLCocoa):
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLCocoa::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLCocoa::setDrawingBufferColorSpace):
(WebCore::GraphicsContextGLCocoa::displayBuffer):
(WebCore::GraphicsContextGLCocoa::drawingBuffer):
(WebCore::GraphicsContextGLCocoa::bindNextDrawingBuffer):
(WebCore::GraphicsContextGLCocoa::freeDrawingBuffers):
(WebCore::GraphicsContextGLCocoa::prepareForDisplay):
(WebCore::GraphicsContextGLCocoa::readCompositedResults):
(WebCore::GraphicsContextGLCocoa::paintCompositedResultsToVideoFrame):
(WebCore::GraphicsContextGLCocoa::withDrawingBufferAsNativeImage):
(WebCore::GraphicsContextGLCocoa::withDisplayBufferAsNativeImage):
(WebCore::GraphicsContextGLCocoa::markDisplayBufferInUse): Deleted.
(WebCore::GraphicsContextGLCocoa::reshapeDisplayBufferBacking): Deleted.
(WebCore::GraphicsContextGLCocoa::allocateAndBindDisplayBufferBacking): Deleted.
(WebCore::GraphicsContextGLCocoa::bindDisplayBufferBacking): Deleted.
(WebCore::GraphicsContextGLANGLE::readCompositedResults): Deleted.
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLIOSurfaceSwapChain.cpp: Removed.
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLIOSurfaceSwapChain.h: Removed.
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.cpp:
(WebCore::GraphicsContextGLFallback::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLFallback::reshapeDisplayBufferBacking): Deleted.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLFallback.h:
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLGBM::reshapeDisplayBufferBacking): Deleted.
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLTextureMapperANGLE::reshapeDrawingBuffer):
(WebCore::GraphicsContextGLTextureMapperANGLE::reshapeDisplayBufferBacking): Deleted.
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::prepareForDisplay):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/263496@main
